### PR TITLE
Keras 1.0 preview

### PR DESCRIPTION
**Pull request by fchollet**
This is a near-total rewriting of the Keras codebase, still undergoing but already 95% complete. It will not be merged into master, rather it will replace master, which means that the merged PRs from the past two weeks will need to be backported.

Some notes:
- this introduces the functional API and deprecates the `Graph` model. See `docs/templates/complete_guide_to_the_keras_functional_api.md` for an (as of yet incomplete) introduction. You can also check out the unit tests `tests/keras/engine/test_topology.py`.
- this introduces modular metrics, including the ability to monitor arbitrary lists of metrics on arbitrary outputs of a multi-output model.
- this introduces support for multi-input and multi-output layers.
- this improves RNNs by providing 4 different implementations to choose from in Theano and 2 different implementations in TensorFlow, each with a different memory/cpu profile. This will allow users to have fast RNNs across very different setups and tasks.
- this improves `Lambda`.
- and much more.

What works at the current time:
- everything, modulo a few things that were overlooked and a few potentially remaining bugs (only one is currently known, fix pending). All examples are running, and the following tests are passing: `test_topology` (functional API test), `test_sequential`, `test_graph`, all integration tests...

TODO:
- backport recent modifications from master into `keras-1`
- rewrite unit tests. In the process of getting them to pass, some bugs will have to be fixed (the set of tests that are passing is already very extensive, so the bugs should be few and minor).
- rewrite documentation.

Get involved:
You can take a look at unit tests.
All further PRs to Keras should be made to the present branch, not to master.

### CODE REVIEW

**Comment by EderSantana** [comment]
can you give us a link to the coveralls again? just to know which parts need unit tests the most.

**Comment by fchollet** [comment]
coveralls stopped working a long time ago. I think we need to pay for it,
but I can't be sure, because the UX of that service is just terrible...

**Comment by fchollet** [comment]
If you want to work on unit testing, you can pursue the rewriting of layers unit tests. So far I have only covered `test_core` (still in progress) and `test_advanced_activations`.

Here's how it works: https://github.com/fchollet/keras/blob/keras-1/tests/keras/layers/test_core.py

Testing RNNs and Embedding will be somewhat more involved though. 

If you decide to do it, please let me know which ones you are doing asap, because I am working on it as well. It would be too bad to do redundant work.

**Comment by fchollet** [comment]
Backporting recently merged PRs into the new version is also a super useful thing to do.

**Comment by EderSantana** [comment]
I'll do some backporting first. I noticed the `batch_dot` I added was lost for example. I'll let you know my next step later once I get more familiar with the changes.

**Comment by fchollet** [comment]
Remaining: 2 tests failing with TF, 5 with Theano. Over 140 passing. It's happening, guys.

**Comment by fchollet** [comment]
All unit tests are new rewritten, and passing, except for one. The last one is `tests/keras/test_sequential_model.py::test_merge_dot`. @EderSantana this is your domain, take a look at it when you have some time.

I think coverage has increased with the rewrite, as well.

We are now very close to the release. Remaining: backporting recent PRs, and rewriting the docs. I will take care of the docs. Anyone wanting to help out should look into PR backporting.

**Comment by EderSantana** [comment]
sorry for being late, I was finishing experiments for a paper... whooaaa! you already did all that? sorry for leaving you alone T_T I'll take a look at my part right now

**Comment by udibr** [comment]
By looking at the code I think there is a bug in stateful shared recurrent object. The same state will be used in the different usages of the same object but the expected functionality is that each usage will have its own separate state. This is different from the weights that are expects to be the same.  However when a recurrent object is stateful you expect it to connect the state across the same input for different batches but each usage of the shard recurrent recives a different input stream.

**Comment by EderSantana** [comment]
@fchollet I got the bug on merge_dot. But how are going to use this? AFAIK merge dot and cos are used to merge 3D tensors resulting in 2D tensors since "dot" and "cost" are reduce-merge. In the test, you merged 2D tensors and resulted in a 1D tensor that Dense layer don't accept. I'll modify the test to cover the first case.

Let me know if I'm wrong and if you want merge-dot to have a different behavior

**Comment by fchollet** [comment]
Deciding what stateful shared layers are supposed to do will require some
thinking. It's definitely not obvious.

I think in general dot is for merging nD tensors into (n-1)D. But dot on 2d
tensors makes perfect sense, only it should return 2d tensors with 1 as the
1st dimension. This will require a ndim check and expand_dims.

**Comment by EderSantana** [comment]
got it!

**Comment by EderSantana** [comment]
Also, we should look into examples next.

I'll work on `babi_memnn` first. I got an error:
```python
Traceback (most recent call last):
  File "examples/babi_memnn.py", line 199, in <module>
    metrics=['accuracy'])
  File "/Users/eder/dev/keras/keras/models.py", line 320, in compile
    **kwargs)
  File "/Users/eder/dev/keras/keras/engine/training.py", line 576, in compile
    updates=train_updates, **kwargs)
  File "/Users/eder/dev/keras/keras/backend/theano_backend.py", line 494, in function
    return Function(inputs, outputs, updates=updates, **kwargs)
  File "/Users/eder/dev/keras/keras/backend/theano_backend.py", line 480, in __init__
    **kwargs)
  File "/Users/eder/anaconda/envs/keras-dev/lib/python2.7/site-packages/Theano-0.8.1-py2.7.egg/theano/compile/function.py", line 320, in function
    output_keys=output_keys)
  File "/Users/eder/anaconda/envs/keras-dev/lib/python2.7/site-packages/Theano-0.8.1-py2.7.egg/theano/compile/pfunc.py", line 442, in pfunc
    no_default_updates=no_default_updates)
  File "/Users/eder/anaconda/envs/keras-dev/lib/python2.7/site-packages/Theano-0.8.1-py2.7.egg/theano/compile/pfunc.py", line 187, in rebuild_collect_shared
    (store_into, update_d[store_into]))
ValueError: ('this shared variable already has an update expression', (embedding_2_W, Elemwise{sub,no_inplace}.0))
```

There are replicate gradient updates.

**Comment by fchollet** [comment]
I tested it a few days ago with Theano and it was working fine. I suppose some change made since would have broke it. Interestingly, it works fine with TensorFlow.

> And this rises another question: where should I document backend functions? I want to avoid having to copy paste from the two backends (and this will avoid copy pastes when we get more backends).

For now, in both backends. We will think of a better solution later. Note that the docs won't be identical for all backends, because some behaviors are backend-specific (even though the API is common). See `K.rnn` for instance.

**Comment by fchollet** [comment]
The failure of babi_memnn is almost certainly due to changes made to the optimizers. The set of weights is correct (non-redundant). The redundant update is an optimization accumulator.

**Comment by fchollet** [comment]
The memnn problem can be solved by ensuring weight unicity at the level of `topology.Container.trainable_weights`. There must be a more elegant solution however.

**Comment by EderSantana** [comment]
can we use something like `set` to filter repeated tensors on `updates`?

**Comment by fchollet** [comment]
A set of pointers has a non-deterministic order, so that would prevent
optimizer state saving/loading. But yes, deduping tensors before getting
updates is an easy fix, and with no side effects. However it would be
preferable to figure out the root cause and fix that.

**Comment by fchollet** [comment]
I have found the root cause, this commit: https://github.com/fchollet/keras/pull/2152/commits/91a819fb343d27e05a5cdf44ac84b43892bdaa56

So now I am confident that deduping weights before updates computation is a good fix.

**Comment by grahamannett** [comment]
> Also, we should look into examples next.
Hey, I'd like to do one of these, any in particular you can think of that would be helpful or just take my pick?

**Comment by fchollet** [comment]
In theory all examples should be working. If you want one that's not
working, feel free to send a PR to fix it.

**Comment by grahamannett** [comment]
Okay, I will look over them. I noticed for example (with addition_rnn.py for example) that slice_X was moved from keras.models to keras.engine.training so possibly its just some simple things that I can help with.

**Comment by fchollet** [comment]
We are almost there, the very last bug is a `lambda` serialization issue with Python 3: https://travis-ci.org/fchollet/keras/builds/120534671

Otherwise, there are still a few recent PRs that we need to backport.

**Comment by fchollet** [comment]
Guys, it's green now. https://travis-ci.org/fchollet/keras/builds/120677181

**Comment by neggert** [comment]
I'm seeing performance issues when building complicated networks again. Specifically, I ported my Inception-v3 architecture to the new functional API and it's taking a long time to construct the model. Haven't profiled, but from the stack trace when I killed it, it looks like all the time is being spent in `build_map_of_graph`.

Can we cache somehow? IIRC, this was the solution to the performance issues when I constructed this model with the Graph API.

**Comment by fchollet** [comment]
@neggert can you post your model as a gist (or even as a script in `examples/`)? That will help with fixing the issue. Yes, I expect that caching nodes or layers should fix the issue, since it sounds like it would be due to complicated recursion patterns in DAGs.

**Comment by fchollet** [comment]
FYI I tried stacking 4 regular inception modules, and the time take by model instantiation (of which `build_map_of_graph` is part) was 0.002s. I can give a shot at reimplementing inceptionv3, but really it would be faster to see your code.

**Comment by neggert** [comment]
Here's the gist: https://gist.github.com/neggert/f8b86d001a367aa7dde1ab6b587246b5

I may have been a bit quick on the trigger with posting. It ends up taking ~15 minutes to construct the model, which is still less than the ~1 hour it takes to compile with Theano. So I'm not sure how much of an issue this is.

I do have some questions about what "depth" means in this model, since many nodes will have different depths along different paths. But I guess as long as it works...

BTW, great work on the functional API. This is so much cleaner than the implementation I have using `Graph`.

**Comment by fchollet** [comment]
> I do have some questions about what "depth" means in this model, since many nodes will have different depths along different paths. But I guess as long as it works...

The depth of a node is the max depth across all paths. Try to draw a few graphs, you'll see it just makes sense.

I have just reimplemented inceptionv3 (without batchnorm though) up until the auxiliary head, and build time is 2s. Total compile time in TF is 20s. So I wonder what's going on there.

Btw I see you implement weight decay with a L2 regularizer. Is it really the same thing?

**Comment by neggert** [comment]
Hmm. I removed batch norm and here's what I get (everything under Theano):

    488.622755051 s to make model
    70.1582930088 s to compile

And here it was with BN:

    893.500264168 s to make model
    503.544885159 s to compile

(It turns out that I either mis-remembered how long it took to compile or keras-1 somehow speeds up the compile)

So BN is a pretty significant contributor. I do think that the number of calls to `build_map_of_graph` will scale with the number of paths, which grows roughly geometrically with the number of layers. So you may actually need to implement the full thing to actually see the problem.

**Comment by fchollet** [comment]
> or keras-1 somehow speeds up the compile

It does, because the training graph and testing graph are now a single graph instead of two separate graphs.

>  Oh, and weight decay and L2 weight regularization are the same thing, right?

Not necessarily. Another way to do weight decay is to multiply weights by `1 - decay` at every update, rather than adding `decay * L2(weights)` to the loss function (which is what the Keras L2 regularizer does). I am wondering what is meant by weight decay in inception v3.

> So you may actually need to implement the full thing to actually see the problem.

I'll try. BTW I pushed some changes to `build_map_of_graph`; although I don't think there should be any significant impact, you can try timing again with the new version.

**Comment by neggert** [comment]
I believe the two formulations of weight decay are equivalent up to some constant factors and maybe order of operations. The extra term in the loss changes the gradient update to

    w_{i + 1} = w_i - LR * (dLoss/dw + decay * w_i) = w_i (1 - LR * decay) - LR * dLoss/dw

Oh, and timing is unchanged with ed4a95b.

**Comment by fchollet** [comment]
Yeah, I got what the problem was. Indeed caching recursive calls solves it. Build time will be sub-1s for you at the next commit.

**Comment by neggert** [comment]
6 seconds without BN, 30 seconds with. Good enough for me. Thanks!

**Comment by fchollet** [comment]
It's pushed. BTW have you ported the pre-trained inception v3 weights to Keras? I was planning on developing an inception v3 script for Keras, but the main point would have been the ability to reuse the ImageNet-trained weights. If you already have that it would be great to add to the repo.

**Comment by fchollet** [comment]
> 6 seconds without BN, 30 seconds with

Maybe something else is taking time, you should try profiling the code. My own implementation went from 6s to 0.007s so I would expect you to see a bigger speed-up.

**Comment by neggert** [comment]
No, I haven't ported the weights. I've been training from scratch on a private dataset.

I'd be happy to contribute the model to the examples, but I can't commit time to porting the weights or training new ones.

**Comment by fchollet** [comment]
I've just updated the documentation (not the one online, the one from the `keras-1` branch). If you guys could check it out and comment on what could be improved or what's missing, I would really appreciate it.

Also: I've made function compilation lazy, i.e. it's only done when a method that will use a specific function is called. I believe this speeds things up a bit.

**Comment by qzshadow** [comment]
Would you introduce the Masking Layer and what is the 'mask_value' in the new doc? Or maybe some examples. I found that many people are confusing(including myself) how to use Masking Layer properly. e.g. #2216 and #2208 etc.
By the way, I found a typo of a file name: keras-1/docs/templates/layers/writing-your-own-keras-lqyers.md, it should be writting-your-own-keras-**layers**.md

File: docs/templates/getting-started/functional-api-guide.md [hunk 1]
```diff
@@ -53,6 +64,8 @@ input_sequences = Input(shape=(20, 784))
 processed_sequences = TimeDistributed(model)(input_sequences)
 ```
 
+-----
+
 ## Multi-input and multi-output models
 
 Here's a good use case for the functional API: models with multiple inputs and outputs. The functional API makes it really easy to manipulate a large number of intertwinned datastreams.
```
Comment (kashif) [inline docs/templates/getting-started/functional-api-guide.md]: intertwinned -> intertwined

File: docs/templates/getting-started/functional-api-guide.md [hunk 2]
```diff
@@ -66,61 +79,85 @@ Here's what our model looks like:
 
 Let's implement it with the functional API.
 
+The main input will receive the headline, as a sequence of integers (each integer encodes a word).
+The integers will be between 1 and 10000 (a vocabuary of 10000 words) and the sequences will be 100 words long.
```
Comment (kashif) [inline docs/templates/getting-started/functional-api-guide.md]: vocabuary -> vocabulary

**Comment by fchollet** [comment]
Updated docs.

**Comment by fchollet** [comment]
We are good to go as far as I can tell. I will keep fixing any issues you can find, but otherwise development is done.

**Comment by dblN** [comment]
I think `K.in_train_phase`(`K.in_test_phase`) acts completely opposite way.
On training, `_LEARNING_PHASE` is true, isn't it? Then, `T.switch(~)` returns `x` not `expression`. 

```
def in_train_phase(x, expression):
    x = T.switch(_LEARNING_PHASE, x, expression)
    x._uses_learning_phase = True
    return x
```

**Comment by fchollet** [comment]
> I think K.in_train_phase(K.in_test_phase) acts completely opposite way. On training, _LEARNING_PHASE is true, isn't it? Then, T.switch(~) returns x not expression.

Yes, that's exactly what is supposed to happen here.

**Comment by fchollet** [comment]
FYI there appears to be a big performance regression in generator methods (`fit_generator` etc). Looking into it.

**Comment by EderSantana** [comment]
This might be related to the generator problem #2225
There seems to be memory overflow somewhere

**Comment by fchollet** [comment]
Actually I figured it out, it's a dumb bug. Fix incoming.

**Comment by dblN** [comment]
> > I think K.in_train_phase(K.in_test_phase) acts completely opposite way. On training, _LEARNING_PHASE is true, isn't it? Then, T.switch(~) returns x not expression.
> 
> Yes, that's exactly what is supposed to happen here.

If so, docs might be wrong.
```python
# tensorflow_backend.py
def in_train_phase(x, expression):
    '''Inserts an expression to be only applied at training time.
    Note that `expression` should have the *same shape* as `x`.
    '''
```
`expression` is not be applied at training time.

**Comment by kashif** [comment]
Now that theano has released 0.8.1 does it make sense to change the readme to use that or perhaps remove the need to install from git?

**Comment by kashif** [comment]
:tada:  cool!

**Pull request merged**